lib/server.js

```diff
@@ -208,6 +209,38 @@ function listenPromise (server, listenOptions) {
   })
 }
 
+// This function used to compile a validation function for http version
+function compileValidateHTTPVersion (options) {
+  let bypass = false
+  // key-value object to store valid http version
+  const obj = {}
+  if (options.serverFactory) {
+    // When serverFactory is passed, we cannot identify how to check http version reliably
+    // So, we should skip the http version check
+    bypass = true
+  }
+  if (options.http2) {
+    // HTTP2 must serve HTTP/2.0
+    obj['2.0'] = true
+    if (options.https && options.https.allowHTTP1 === true) {
+      // HTTP2 with HTTPS.allowHTTP1 allow fallback to HTTP/1.1 and HTTP/1.0
+      obj['1.1'] = true
+      obj['1.0'] = true
+    }
+  } else {
+    // HTTP must server HTTP/1.1 and HTTP/1.0
+    obj['1.1'] = true
+    obj['1.0'] = true
+  }
+  // The compiled function here placed in one of the hottest path inside fastify
+  // the implementation here must be as performant as possible
+  return function validateHTTPVersion (httpVersion) {
+    // `bypass` skip the check when custom server factory provided
+    // `httpVersion in obj` check for the valid http version we should support
+    return bypass || httpVersion in obj
```

Comment:
What would the difference be in using an actual `Map` and then a `.has` check?

```suggestion
    return bypass || obj.has(httpVersion)
```

Reply:
`Map` is definitely slower in this case.
I also tested `Set`, it is about the same as `Array.includes`.